astropy/modeling/tests/test_compound.py

```diff
@@ -543,175 +277,77 @@ def test_identity_input():
     model = ident1 & shift | rotation
     assert_allclose(model(1, 2), [-3.0, 1.0])
 
-    # Same test case but using class composition
-    TestModel = ident1 & Shift | Rotation2D
-    model = TestModel(offset_1=1, angle_2=90)
-    assert_allclose(model(1, 2), [-3.0, 1.0])
-
 
-def test_slicing_on_instances_2():
+def test_invalid_operands():
     """
-    More slicing tests.
-
-    Regression test for https://github.com/embray/astropy/pull/10
+    Test that certain operators do not work with models whose inputs/outputs do
+    not match up correctly.
     """
 
-    model_a = Shift(1, name='a')
-    model_b = Shift(2, name='b')
-    model_c = Rotation2D(3, name='c')
-    model_d = Scale(2, name='d')
-    model_e = Scale(3, name='e')
-
-    m = (model_a & model_b) | model_c | (model_d & model_e)
-
-    with pytest.raises(ModelDefinitionError):
-        # The slice can't actually be taken since the resulting model cannot be
-        # evaluated
-        assert m[1:].submodel_names == ('b', 'c', 'd', 'e')
-
-    assert m[:].submodel_names == ('a', 'b', 'c', 'd', 'e')
-    assert m['a':].submodel_names == ('a', 'b', 'c', 'd', 'e')
-
-    with pytest.raises(ModelDefinitionError):
-        assert m['c':'d'].submodel_names == ('c', 'd')
-
-    assert m[1:2].name == 'b'
-    assert m[2:7].submodel_names == ('c', 'd', 'e')
-    with pytest.raises(IndexError):
-        m['x']
-    with pytest.raises(IndexError):
-        m['a': 'r']
-
     with pytest.raises(ModelDefinitionError):
-        assert m[-4:4].submodel_names == ('b', 'c', 'd')
+        Rotation2D(90) | Gaussian1D(1, 0, 0.1)
 
     with pytest.raises(ModelDefinitionError):
-        assert m[-4:-2].submodel_names == ('b', 'c')
-
-
-def test_slicing_on_instances_3():
```

Comment:
Just curious, why were these tests removed? They operate on compound model instances, right?

Reply:
These were all removed for various reasons, but proper replacement tests are lacking and should be provided. One thing that the changes document mentions (and this is subject to further discussion and work) is that compound model slicing works differently now, and is more restrictive. I'll argue that the current slicing behavior can be very surprising since it tolerates slices that would otherwise be considered poorly defined. For example if one defined the following compound model (using all 1 input models):

m = m1 * m2 + m3

should the slice [1:] be permitted? In the previous modeling  it is and is equivalent to m2 + m3. And yet, it doesn't seem that it should be permitted. The current implementation requires the slice to be a subtree, which is admittedly over restrictive. For example if

m = m1 + m2 + m3 + m4

m[:3] is permitted, but m[1:] is not since m1 is a member of all subtrees. But to implement both a flexible and correct slicing mechanism is a fair amount of work and we didn't want to delay merging on this.

The question is whether the slicing tests should be added before or after the merge.